### ISSUE TITLE
TabView tabs with an '&' in the title are cut off

### ISSUE THREAD

**Issue by daniels220** [dolphinsmalltalk/Dolphin#989]
Generally, the tabs of a TabView[XP] auto-resize to fit the text they contain, but if that text contains an `&` (which must be escaped as `&&`), the width of the `&` is not being counted, leading to the tab being slightly too narrow and the text being cut off with an ellipsis. I tracked the issue down to `TabViewXP>>paint[Vertical]Label:on:in:`. Looks like back in 2018 (167ee127d) it was changed to do a `copyWithout: $&`, which is perhaps _more_ correct than doing nothing—it works for cases like `Hello &World`—but causes this problem with `&&`.

To get it right we need to scan through the string, removing ampersands but then protecting the immediately-following character, including if it is another ampersand. Easy enough, and I'm happy to do it, but—where would you suggest putting the method to do this? I looked around for other places that deal with this pattern, and found them in five different classes:

* `MenuPainter>>widthOfItemText:using:`
* `Menu>>find:ifAbsent:` (removing only those ampersands that won't be displayed is arguably more correct than removing all of them before comparing—or, if we want to be fuzzy, they should be removed from the search string as well, otherwise searching for `'a & b'` won't find a MenuItem with description `' a && b'`, because it'll be comparing against `'a  b'`).
* `OwnerDrawnButton>>unadornedText` (which, also, perhaps this should be pushed up to `CommandButton`? Or even higher?).
* The abovementioned `TabViewXP` methods.
* `StaticText>>calculateExtent:` (though, right now the `copyWithout: $&` is actually dead code—missing an assignment—and should be reviewed to make sure it's actually correct).

Given the widespread usage, it has to be either a method of String (but, what to call it?), or a class method somewhere (but, where?). Any thoughts?

**Comment by blairmcg** [comment]
I don't think the solution to these related issues should involve string manipulation, rather it requires correcting the underlying API usage.

> * `MenuPainter>>widthOfItemText:using:`

This paints the text using Canvas>>formatText:in:flags:, which wraps the Windows User [DrawTextEx()](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-drawtextexw) API. Therefore for correctness it should really be using that API with the `DT_CALCRECT` flag to calculate the extent, not the Windows GDI [GetTextExtentPoint32 ](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-gettextextentpoint32w) function. 

There is an existing wrapper Canvas>>textExtent:width:alignment: that is used to calculate the extent of multiline text within a specified width. This won't give the correct result for a single line (needs the DT_SINGLELINE flag). Some thought and experimentation is needed as to the design of a new/replacement Canvas API for this.

> * `Menu>>find:ifAbsent:` (removing only those ampersands that won't be displayed is arguably more correct than removing all of them before comparing—or, if we want to be fuzzy, they should be removed from the search string as well, otherwise searching for `'a & b'` won't find a MenuItem with description `' a && b'`, because it'll be comparing against `'a  b'`).

This and the related/dependent methods should really be deprecated. Searching a menu by displayed text for programmatic purposes is not a good practice.

> * `OwnerDrawnButton>>unadornedText` (which, also, perhaps this should be pushed up to `CommandButton`? Or even higher?).

This is part of a contribution that is using it for it's own purposes. It looks like that is mostly because of the same issue with the Windows API used calculating the extent required for the text, except for one case. I'm not sure why there would be a general need for it, but I like the terminology.

> * The abovementioned `TabViewXP` methods.

This is the same as the `MenuPainter`, in that it should be using the correct API to calculate the extent.

> * `StaticText>>calculateExtent:` (though, right now the `copyWithout: $&` is actually dead code—missing an assignment—and should be reviewed to make sure it's actually correct).

Again this should use the DrawTextEx API to calculate the extent for single line form, as it currently does now for multi-line forms.

**Comment by blairmcg** [comment]
Here is a version of the (rather overlong) TabViewXP method with the issue fixed in place.

```smalltalk
paintLabel: anInteger on: aCanvas in: aRectangle
	| inverted rcItem textExtent labelRect textOffset text tcItem imageIndex |
	tcItem := self getItem: anInteger.
	text := tcItem text.
	imageIndex := self hasIcons ifTrue: [tcItem iImage + 1] ifFalse: [-1].
	inverted := self isInverted.
	rcItem := aRectangle copy.
	inverted ifTrue: [rcItem bottom: rcItem bottom - 1] ifFalse: [rcItem top: rcItem top + 1].
	labelRect := RECT newBuffer.
	aCanvas
		formatText: text
		in: labelRect
		flags: ##(DT_SINGLELINE | DT_CALCRECT).
	textExtent := labelRect extent.
	imageIndex <= 0
		ifTrue: 
			[textOffset := 0.
			labelRect := Rectangle center: rcItem center extent: textExtent]
		ifFalse: 
			[| space imageExtent |
			imageExtent := self imageExtent.
			textOffset := imageExtent x + self imageTextGap.
			labelRect := Rectangle center: rcItem center
						extent: (textExtent isZero
								ifTrue: [imageExtent]
								ifFalse: [(textOffset + textExtent x) @ (imageExtent y max: textExtent y + 1)]).
			space := labelRect height - imageExtent y.
			(self imageManager imageListWithExtent: imageExtent)
				draw: imageIndex
				on: aCanvas
				at: labelRect left
						@ (labelRect top + (inverted ifTrue: [space // 2] ifFalse: [space - (space // 2)]))].
	textExtent > 0 ifFalse: [^self].
	labelRect := labelRect insetOriginBy: textOffset @ (inverted ifTrue: [-1] ifFalse: [1]).
	aCanvas
		formatText: text
		in: labelRect
		flags: ##(DT_VCENTER | DT_SINGLELINE | DT_END_ELLIPSIS)
```

**Comment by blairmcg** [comment]
Fixed by c2d6825 and preceding commits listed above.
